theano/tensor/nnet/bn.py

```diff
@@ -0,0 +1,56 @@
+import theano
+from theano.scalar import Composite
+from theano.scalar import add, sub, true_div, mul
+
+
+class BNComposite(Composite):
+
+    def __init__(self, dtype):
+        x = theano.scalar.Scalar(dtype=dtype).make_variable()
+        mean = theano.scalar.Scalar(dtype=dtype).make_variable()
+        var = theano.scalar.Scalar(dtype=dtype).make_variable()
+        gamma = theano.scalar.Scalar(dtype=dtype).make_variable()
+        beta = theano.scalar.Scalar(dtype=dtype).make_variable()
+        o = add(mul(true_div(sub(x, mean),  var), gamma), beta)
+        inputs = [x, mean, var, gamma, beta]
+        outputs= [o]
+        super(BNComposite, self).__init__(inputs, outputs)
+
+    def grad(self, inps, grads):
+        x, mean, var, gamma, beta = inps
+        top, = grads
+        dx = (top*gamma) / var
+        dmean = -(top*gamma) / var
+        dvar = -(top * gamma * (x - mean)) / (var*var)
+        dgamma = top*(x - mean) / var
+        return [dx, dmean, dvar, dgamma, top]
+
+
+def batch_normalization(inputs, gamma, beta, mean, variance, axis=0):
+    """
+    This function will build the symbolic graph for applying batch normalization
+    to a set of activations.
+
+    Parameters
+    ----------
+    inputs : symbolic tensor
+        Mini-batch of examples
+    gamma: symbolic vector
+        BN scale parameter, must be of same dimension that
+        the number of inputs channel
+    beta: symbolic vector
+        BN shift parameter, must be of same dimension that
+        the number of inputs channel
+    mean: symbolic tensor
+        inputs means
+    variance: symbolic tensor
+        inputs variance
+    axis: int
+        channel axis
```

Comment:
Ah wait, `axis` is not used at all, since you're requiring inputs to be broadcasted outside of that function. So you can either remove the `axis` argument completely, or move the broadcasting inside here, and allow `axis` to be a list, something like:
```python
if isinstance(axis, int):
    axis = [axis]
dims = range(gamma.ndim)
pattern = [next(dims) if dim in axis else 'x' for dim in range(inputs.ndim)]
gamma = gamma.dimshuffle(*pattern)
beta = beta.dimshuffle(*pattern)
mean = mean.dimshuffle(*pattern)
variance = variance.dimshuffle(*pattern)
```

But I think ultimately it gives a more flexible implementation if you just drop the `axis` argument and require everything to be broadcastable against `inputs` (i.e., just alter the docstring, keep the implementation as it is).